### PR TITLE
(DOCSP-43745) Connect to MongoDB

### PR DESCRIPTION
**Pull request by amalhotra-mdb**
# Pull Request Info

Fill out root page for "Connect to MongoDB" (created as stub in [this ticket](http://example.com/)).

A/C: I can see all the operations for connecting to MongoDB in a section.

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - <https://jira.mongodb.org/browse/DOCSP-43745>
Staging - <https://deploy-preview-61--docs-c.netlify.app/connect/>

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?
- [ ] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/includes/usage-examples/connect-sample-app.c [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  // End example code here
+
+  database = mongoc_client_get_database (client, "admin");
+    
+  ping = BCON_NEW ("ping", BCON_INT32 (1));
+   
+  if (!mongoc_client_command_simple (client, "admin", ping, NULL, &reply, &error)) {
+	  fprintf (stderr, "%s\n", error.message);
+    goto cleanup;
```
Comment (mcmorisi) [inline source/includes/usage-examples/connect-sample-app.c:22]: Q: Should these have the same indentation?

File: source/connect.txt [hunk 2]
```diff
@@ -28,6 +28,182 @@ Connect to MongoDB
...
+Replica Set
+~~~~~~~~~~~
+
+The following code shows the connection string to connect to a
+replica set:
+
+.. code-block:: c
+   
+   char *uri = "mongodb://<replica-set-member>:<port>/?replicaSet=<replica-set-name>";
+   mongoc_client_t *client = mongoc_client_new (uri);
```
Comment (mcmorisi) [inline source/connect.txt:104]: Nit: we use underlines for the placeholders elsewhere in this page, I think we should stick to that.

```suggestion
   char *uri = "mongodb://<replica_set_member>:<port>/?replicaSet=<replica_set_name>";
   mongoc_client_t *client = mongoc_client_new (uri);
```

File: source/includes/usage-examples/connect-sample-app.c (hunk 1 above)
Comment (amalhotra-mdb) [reply source/includes/usage-examples/connect-sample-app.c]: yup fixed it!

File: source/includes/usage-examples/connect-sample-app.c [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#include <mongoc/mongoc.h>
+#include <bson/bson.h>
+
+int main(void) {
+
+  mongoc_database_t *database
+  bson_t *ping, reply;
```
Comment (kevinAlbs) [inline source/includes/usage-examples/connect-sample-app.c:7]:
```suggestion
  mongoc_database_t *database = NULL;
  bson_t *ping = NULL, reply = BSON_INITIALIZER;
```

Add missing semicolon. Suggest initializing values so `goto cleanup;` is safe to call anywhere after.

Example:

```c
bson_t *ping;

if (something_may_fail()) {
   goto cleanup; // Destroys uninitialized `ping`!
}

ping = BCON_NEW ("ping", BCON_INT32(1));

cleanup:
   bson_destroy (ping);
```

File: source/includes/usage-examples/connect-sample-app.c [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  bson_error_t error;
+
+  mongoc_init();
+
+  // Start example code here
+
+  // End example code here
```
Comment (kevinAlbs) [inline source/includes/usage-examples/connect-sample-app.c:14]: Copying an example directly destroys the client before the database operation. Example from `tls-tabs.rst`:

```c
mongoc_client_t *client = mongoc_client_new("mongodb://<db_username>:<db_password>@<hostname>/?tls=true");
// Do database work here
mongoc_client_destroy (client);
```

Suggest adding declaration of `client`/`uri` into this file:

```c
#include <mongoc/mongoc.h>
#include <bson/bson.h>

int main(void) {
  mongoc_uri_t* uri = NULL;
  mongoc_client_t *client = NULL;
  mongoc_database_t *database = NULL;
  bson_t *ping = NULL, reply = BSON_INITIALIZER;
  bson_error_t error;

  mongoc_init();

  // Start example code here

  // End example code here

  database = mongoc_client_get_database (client, "admin");
    
  ping = BCON_NEW ("ping", BCON_INT32 (1));
   
  if (!mongoc_client_command_simple (client, "admin", ping, NULL, &reply, &error)) {
    fprintf (stderr, "%s\n", error.message);
    goto cleanup;
  }
  printf ("Pinged your deployment. You successfully connected to MongoDB!\n");

  cleanup:
    bson_destroy (&reply);
    bson_destroy (ping);
    mongoc_database_destroy (database);
    mongoc_client_destroy (client);
    mongoc_uri_destroy (uri);
    mongoc_cleanup ();
}
```

Then revising the examples to assign:
```c
client = mongoc_client_new("mongodb://<db_username>:<db_password>@<hostname>/?tls=true");
// Do database work here
```

**Comment by amalhotra-mdb** [comment]
@kevinAlbs thanks so much for all your help on these C standardization PRs!

File: source/connect.txt [hunk 5]
```diff
@@ -26,8 +26,169 @@ Connect to MongoDB
...
+
+.. code-block:: c 
+
+   bson_error_t error;
+
+   client = mongoc_client_new ("mongodb+srv://<db_username>:<db_password>@<hostname>/?<options>");
+
+   // Set the version of the Stable API on the client
+   mongoc_server_api_t *api = mongoc_server_api_new(MONGOC_SERVER_API_V1);
+   mongoc_client_set_server_api(client, api, &error)
```
Comment (kevinAlbs) [inline source/connect.txt:187]:
```suggestion
   mongoc_client_set_server_api(client, api, &error);
```

File: source/connect.txt [hunk 6]
```diff
@@ -26,8 +26,169 @@ Connect to MongoDB
...
+Local Deployment
+~~~~~~~~~~~~~~~~
+
+The following code shows the connection string to connect to a local
+instance of MongoDB:
+
+.. code-block:: c
+
+   uri = mongoc_uri_new ("mongodb://localhost:27017"); 
+   client = mongoc_client_new_from_uri (uri);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply source/connect.txt:73]: Not yet addressed.

**Review by kevinAlbs — APPROVED** [review]
LGTM with one comment addressed.

**Pull request merged**